### PR TITLE
Feature: Enable search in grid/wall view

### PR DESCRIPTION
**Pull request by wutschel**
## Description

Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/467.

This PR enables the search in grid/wall view (tested with albums, artists and movies):
- Search field usable as for list view
- Correct results are shown
- Longpressing or selecting an item processes the details of the correct item
- Show amount of found items in section header

Screenshots:
https://abload.de/img/simulatorscreenshot-iqrkga.png
https://abload.de/img/simulatorscreenshot-iajj2g.png
https://abload.de/img/simulatorscreenshot-i9tk5e.png

## Summary for release notes

Feature: Enable search in grid/wall view

### CODE REVIEW

**Comment by wutschel** [comment]
Should be merged after #521.

**Comment by kambala-decapitator** [comment]
Please fix conflicts

**Comment by wutschel** [comment]
Rebased to master.

File: XBMC Remote/DetailViewController.m [hunk 1]
```diff
@@ -1420,15 +1425,23 @@ - (UICollectionReusableView*)collectionView:(UICollectionView*)cView viewForSupp
...
     if (episodesView) {
         return ([self.sectionArrayOpen[section] boolValue] ? [[self.sections objectForKey:self.sectionArray[section]] count] : 0);
     }
     return [[self.sections objectForKey:self.sectionArray[section]] count];
 }
 
 - (UICollectionViewCell*)collectionView:(UICollectionView*)cView cellForItemAtIndexPath:(NSIndexPath*)indexPath {
-    
-    NSDictionary *item = [self.sections objectForKey:self.sectionArray[indexPath.section]][indexPath.row];
+    NSDictionary *item = nil;
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m]: FYI under ARC `nil` is assigned to objects implicitly

File: XBMC Remote/DetailViewController.m [hunk 2]
```diff
@@ -1420,15 +1425,23 @@ - (UICollectionReusableView*)collectionView:(UICollectionView*)cView viewForSupp
...
+    if ([self doesShowSearchResults]) {
+        item = self.filteredListContent[indexPath.row];
+    }
+    else {
+        item = [self.sections objectForKey:self.sectionArray[indexPath.section]][indexPath.row];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m]: I suggest to extract this into a separate function (or maybe even the whole if/else block)

File: XBMC Remote/DetailViewController.m [hunk 3]
```diff
@@ -1398,9 +1398,11 @@ - (void)initCollectionView {
 
 - (NSInteger)numberOfSectionsInCollectionView:(UICollectionView*)collectionView {
     if ([self doesShowSearchResults]) {
+        self.indexView.hidden = YES;
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:1429]: I don't find it a good place for this logic: should be done at the point where you toggle between search and normal view

File: XBMC Remote/DetailViewController.m [hunk 4]
```diff
@@ -426,6 +426,20 @@ - (void)updateSyncDate:(NSString*)filePath {
...
+- (NSString*)getAmountOfSearchResultsString {
+    NSString *results = @"";
+    int numResult = (int)self.filteredListContent.count;
+    if (numResult) {
+        if (numResult != 1) {
+            results = [NSString stringWithFormat:LOCALIZED_STR(@"%d results"), numResult];
+        }
+        else {
+            results = LOCALIZED_STR(@"1 result");
+        }
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:447]: now it's time to start localizing plurals correctly ;)

File: XBMC Remote/DetailViewController.m [hunk 5]
```diff
@@ -850,6 +850,10 @@ - (void)configureLibraryView {
         if (self.indexView.indexTitles.count > 1) {
             self.indexView.hidden = NO;
         }
+        // Need to remove the searchController frm the dataList header view. Otherwise the search
+        // will not  correctly show on top of the grid view.
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:853]: please fix a couple of typos :)

File: XBMC Remote/DetailViewController.m [hunk 6]
```diff
@@ -5449,6 +5446,21 @@ - (void)searchBarTextDidBeginEditing:(UISearchBar*)searchbar {
 
 #pragma mark UISearchController Delegate Methods
 
+- (void)initSearchController {
+    self.searchController = [[UISearchController alloc]initWithSearchResultsController:nil];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:5403]:
```suggestion
    self.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
```

File: XBMC Remote/DetailViewController.m [hunk 7]
```diff
@@ -5628,8 +5629,6 @@ - (void)viewDidLoad {
         dataList.separatorInset = UIEdgeInsetsZero;
         dataList.separatorColor = [Utilities getGrayColor:38 alpha:1];
     }
-    self.searchController.searchBar.tintColor = [Utilities get2ndLabelColor];
-    self.searchController.searchBar.backgroundColor = [Utilities getSystemGray6];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:5612]: is it correct to remove these lines?

File: XBMC Remote/DetailViewController.m [hunk 8]
```diff
@@ -2858,55 +2858,6 @@ - (CGFloat)tableView:(UITableView*)tableView heightForFooterInSection:(NSInteger
     return 0;
 }
 
-#pragma mark - ScrollView Delegate
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:2838]: I guess declaration of `UIScrollViewDelegate` conformance can be removed as well?

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m [hunk 9]
```diff
@@ -171,8 +171,21 @@ - (void)buildIndexLabels {
         label.font = [UIFont boldSystemFontOfSize:11];
         label.minimumScaleFactor = 11.0/11.0;
         label.adjustsFontSizeToFitWidth = YES;
+        if ([indexTitle isEqualToString:@"🔍"]) {
+            // Use the same icon as iOS's index (magnifying glass with transparent circle)
+            UIImage *searchImage = [UIImage imageNamed:@"icon_search"];
+            searchImage = [Utilities colorizeImage:searchImage withColor:UIColor.systemBlueColor];
+            UIImageView *imageView = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, self.frame.size.width, 11)];
```
Comment (kambala-decapitator) [inline XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:178]:
```suggestion
            UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, 11)];
```

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m [hunk 10]
```diff
@@ -171,8 +171,21 @@ - (void)buildIndexLabels {
...
+            imageView.contentMode = UIViewContentModeScaleAspectFit;
+            imageView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin;
+            imageView.image = searchImage;
+            [label addSubview:imageView];
```
Comment (kambala-decapitator) [inline XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:182]:
```suggestion
            imageView.image = searchImage;

            [label addSubview:imageView];
```

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m [hunk 11]
```diff
@@ -171,8 +171,21 @@ - (void)buildIndexLabels {
         label.font = [UIFont boldSystemFontOfSize:11];
         label.minimumScaleFactor = 11.0/11.0;
         label.adjustsFontSizeToFitWidth = YES;
+        if ([indexTitle isEqualToString:@"🔍"]) {
+            // Use the same icon as iOS's index (magnifying glass with transparent circle)
```
Comment (kambala-decapitator) [inline XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:175]: I can suggest to use iOS system image on iOS 14+, you can find the list of system images in SF Symbols app

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/DetailViewController.m (hunk 4 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:447]: Already created https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/479 for this whole thing. :)

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m (hunk 11 above)
Comment (wutschel) [reply XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:175]: I thought of the same, but dropped this as I wanted to have the same icons for all iOS versions. As long as we keep supporting older iOS version I would to keep it like this.

Comment (kambala-decapitator) [reply XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:175]: Personally I disagree on this, but not insisting on using system images

File: XBMC Remote/DetailViewController.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: This is removed with a later commit.

File: XBMC Remote/DetailViewController.m (hunk 5 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:853]: Done

File: XBMC Remote/DetailViewController.m (hunk 8 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:2838]: Interesting. The delegate conformance was removed even before this commit about 9 years back (https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/0c53255b9f83f0f65b3800e32936c3cd9280bb15).

File: XBMC Remote/DetailViewController.m (hunk 6 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:5403]: Done.

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m (hunk 9 above)
Comment (wutschel) [reply XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:178]: Done.

File: XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m (hunk 10 above)
Comment (wutschel) [reply XBMC Remote/BDKCollectionIndexView/BDKCollectionIndexView.m:182]: Done.

File: XBMC Remote/DetailViewController.m (hunk 7 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:5612]: Yes, these colors are set in `configureLibraryView`. The removed lines were obsolete.

File: XBMC Remote/DetailViewController.m (hunk 3 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: This took a while. You again hit the right spot. The visibility of the grid view's index was not handled well at all. I now reworked this. In addition i came across a few more issues with iPad's fullscreen view and found solutions (in 1 case a workaround which will do for now).

File: XBMC Remote/DetailViewController.m [hunk 12]
```diff
@@ -426,6 +426,28 @@ - (void)updateSyncDate:(NSString*)filePath {
 
 #pragma mark - Utility
 
+- (void)setCollectionViewIndexVisibility {
+    if (enableCollectionView) {
+        // Get the index titles
+        NSMutableArray *tmpArr = [[NSMutableArray alloc] initWithArray:self.sectionArray];
+        if (tmpArr.count > 1) {
+            [tmpArr replaceObjectAtIndex:0 withObject:[NSString stringWithUTF8String:"\xF0\x9F\x94\x8D"]];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:434]: what is this? Please use the UTF-8 string in code directly.

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/DetailViewController.m (hunk 12 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: This is the magnifying lens, see the updated commit.

**Review by kambala-decapitator — APPROVED** [review]

**Comment by wutschel** [comment]
Squashed fixups and rebased to master.

**Pull request merged**